src/hitchhiker/sqlite.clj

```diff
@@ -0,0 +1,235 @@
+            [clojure.core.cache :as cache]
+            [taoensso.nippy :as nippy])
+  (:import [java.sql SQLException]))
+
+(defn underscore [x]
+  (str/replace (str x) "-" "_"))
+
+(def schema
+  {:hh-key (jdbc/create-table-ddl :hh-key
+             [[:k :string]
+              [:v :blob]
+              [:rc :integer "default 0"]]
+             {:entities underscore})
+
+   :hh-ref (jdbc/create-table-ddl :hh-ref
+             [[:parent :string "references hh_key(k) on delete cascade"]
+              [:child :string "references hh_key(k) on delete cascade"]]
+             {:entities underscore})
+
+   :hh-ref-by-parent "create index if not exists hh_ref_by_parent on hh_ref (parent);"
+   :hh-ref-by-child "create index if not exists hh_ref_by_child on hh_ref (child);"})
+
+(def query
+  {:table-exists? "select 1 from sqlite_master where type='table' and name=?"
+   :index-exists? "select 1 from sqlite_master where type='index' and name=?"
+   :find-key      "select * from hh_key where k=?"})
+
+(defn drop-ref [db key]
+  (jdbc/delete! db :hh-key ["k = ?" key]))
+
+(defn db-spec [subname]
+  {:classname   "org.sqlite.JDBC"
+   :subprotocol "sqlite"
+   :subname     subname})
+
+(defn table-exists? [db table]
+  (-> [(query :table-exists?) (underscore table)]
+      (jdbc/query db)
+      (not-empty)))
+
+(defn index-exists? [db idx]
+  (-> [(query :index-exists?) (underscore idx)]
+      (jdbc/query db)
+      (not-empty)))
+
+(defn create-table [db tbl]
+  (try
+    (jdbc/execute! db (:ddl tbl))
+    (catch SQLException e
+      (when-not (re-matches #"table (.*) already exists" (.getMessage e))
+        (throw (ex-info "failed to create table"
+                        {:tbl tbl
+                         :ddl (:ddl tbl)}
+                        e))))))
+
+(defn create-index [db idx]
+  (try
+    (jdbc/execute! db (:ddl idx))
+    (catch SQLException e
+      (when-not (re-matches #"index (.*) already exists" (.getMessage e))
+        (throw (ex-info "failed to create index"
+                        {:idx idx}
+                        e))))))
+
+(defn ensure-schema [db]
+  (let [ensure (fn [{:keys [schema exists? create!]}]
+                 (doseq [item schema]
+                   (when-not (exists? db (underscore (name item)))
+                     (create! db (schema item)))))]
+
+    (ensure {:schema [:hh-key :hh-ref]
+             :exists? table-exists?
+             :create! create-table})
+
+    (ensure {:schema [:hh-ref-by-parent :hh-ref-by-child]
+             :exists index-exists?
+             :create! create-index})))
+
+(defn add-node [db {:keys [k v] :as node}]
+  (try
+    (jdbc/insert! db :hh-key {:k k :v (nippy/freeze v)} {:entities hyphenate})
+    (catch SQLException e
+      (throw (ex-info "failed to add node"
+                      {:node node
+                       :db db} e)))))
+
+(defn delete-key [db k]
+  (jdbc/delete! :hh-key ["k = ?" k]))
+
+(defn add-refs [db {:keys [parent children]}]
```

Comment:
Great! Once you're able to run the benchmark, let's jump on a hangout and I'll walk you through some of the techniques I use to identify perf bottlenecks--we'll look for operations that use too much CPU, too much I/O, or cause weird cross-interferences :). You should have my google account from invite I requested on drive.